presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/WindowFilterPushDown.java

```diff
@@ -60,220 +64,154 @@ public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types
         }
 
         @Override
-        public PlanNode visitLimit(LimitNode node, RewriteContext<Constraint> context)
+        public PlanNode visitLimit(LimitNode node, RewriteContext<Void> context)
         {
             // Operators can handle MAX_VALUE rows per page, so do not optimize if count is greater than this value
             if (node.getCount() >= Integer.MAX_VALUE) {
                 return context.defaultRewrite(node);
             }
-            Constraint constraint = new Constraint(Optional.of((int) node.getCount()), Optional.empty());
-            PlanNode rewrittenSource = context.rewrite(node.getSource(), constraint);
 
-            if (rewrittenSource != node.getSource()) {
-                return new LimitNode(idAllocator.getNextId(), rewrittenSource, node.getCount());
+            PlanNode source = context.rewrite(node.getSource());
+            if (source instanceof RowNumberNode) {
+                RowNumberNode rowNumberNode = (RowNumberNode) source;
+                int limit = Ints.checkedCast(node.getCount());
+                if (rowNumberNode.getMaxRowCountPerPartition().isPresent()) {
+                    limit = Math.min(rowNumberNode.getMaxRowCountPerPartition().get(), limit);
+                }
+                source = rowNumberNode.withLimit(Ints.checkedCast(limit));
+            }
+            else if (source instanceof WindowNode) {
+                WindowNode windowNode = (WindowNode) source;
+                if (canOptimizeWindowFunction(windowNode)) {
+                    // verify that unordered row_number window functions are replaced by RowNumberNode
+                    verify(!windowNode.getOrderBy().isEmpty());
+                    source = convertToTopNRowNumber(windowNode, Ints.checkedCast(node.getCount()));
+                }
             }
-
-            return context.defaultRewrite(node);
+            return context.replaceChildren(node, ImmutableList.of(source));
         }
 
         @Override
-        public PlanNode visitFilter(FilterNode node, RewriteContext<Constraint> context)
+        public PlanNode visitFilter(FilterNode node, RewriteContext<Void> context)
         {
-            PlanNode rewrittenSource = context.rewrite(node.getSource(), new Constraint(Optional.empty(), Optional.of(node.getPredicate())));
-            if (rewrittenSource != node.getSource()) {
-                if (rewrittenSource instanceof TopNRowNumberNode) {
-                    return rewrittenSource;
-                }
-                return new FilterNode(idAllocator.getNextId(), rewrittenSource, node.getPredicate());
-            }
-            return context.defaultRewrite(node);
-        }
-    }
+            PlanNode source = context.rewrite(node.getSource());
 
-    private static boolean isRowNumberSignature(Signature signature)
-    {
-        return signature.equals(ROW_NUMBER_SIGNATURE);
-    }
+            if (source instanceof RowNumberNode || source instanceof WindowNode) {
```

Comment:
you can probably get by without this extra if condition, just make the below if-conditions check for each of these sources separately.